tooling/release_download_test.sh

```diff
@@ -8,164 +8,490 @@
+#
+# A utility function to print verbose output.
+#
+########################################################################################################################
+print_verbose() {
+  if [ "$VERBOSE" = "true" ]; then
+    echo "${BOLD}$(date +%T) : $*${NORMAL}" 1>&2;
+  fi
+}
+
+########################################################################################################################
+#
+# Extract JDK major version from a specified tag.
+#
+########################################################################################################################
+extract_major_version() {
+  if echo "${TAG}" | grep jdk8u > /dev/null; then
+    MAJOR_VERSION=8
+  elif echo "${TAG}" | grep ^jdk- > /dev/null; then
+    MAJOR_VERSION=$(echo "${TAG}" | cut -d- -f2 | cut -d. -f1 | cut -d\+ -f1)
+  else
+    # Probably a beta with the tag starting jdkXXu
+    MAJOR_VERSION=$(echo "${TAG}" | cut -d- -f1 | tr -d jdku)
+  fi
+}
+
+########################################################################################################################
+#
+# Download release information from GitHub for the specified major version.
+# return : file containing release information
+#
+########################################################################################################################
+download_jdk_releases() {
+  local output_file
+  output_file="${WORKSPACE}/jdk${MAJOR_VERSION}.txt"
+
+  if ! curl -sS "https://api.github.com/repos/adoptium/temurin${MAJOR_VERSION}-binaries/releases" > "${output_file}"; then
+     print_error "GitHub API call failed - aborting"
+     exit 2
+  fi
+
+  echo "${output_file}"
+}
+
+########################################################################################################################
+#
+# Download all files of the specified release if required.
+# param 1: jdk release info file
+#
+########################################################################################################################
+download_release_files() {
+  local jdk_releases filter url
 
-[ "$VERBOSE" = "true" ] && echo "$(date +%T) : IVT: Downloading files from release repository"
+  jdk_releases=$1
+
+  cd "${WORKSPACE}/staging/${TAG}" || exit 1
 
-# Leaving this "if/fi" commented out as it can be useful if doing standalone
-# testing to avoid having to re-download. May be removed in future
-#if [ ! -r "staging/$TAG/OpenJDK21U-debugimage_aarch64_mac_hotspot_ea_21-0-35.tar.gz.json" ]; then
-  rm -rf staging
-  mkdir staging "staging/$TAG"
-  cd "staging/$TAG" || exit 3
   # Early access versions are currently in a different format
-  if echo "$TAG" | grep ea-beta; then
-     FILTER="ea_${MAJOR_VERSION}"
+  if echo "${TAG}" | grep ea-beta; then
+    filter="ea_${MAJOR_VERSION}"
   else
-     FILTER=$(echo "$TAG" | sed 's/+/%2B/g')
+    # shellcheck disable=SC2001
+    filter=$(echo "${TAG}" | sed 's/+/%2B/g')
```

Comment:
This is needed to avoid problems when we produce point releases
```suggestion
    filter=$(echo "/${TAG}/" | sed 's/+/%2B/g')
```